### ISSUE TITLE
Jobs stop running on connection error

### ISSUE THREAD

**Issue by gonzoinc** [jameswlane/status-board#188]
- `status-board` version: 1.1.37
- `node` version: v9.8.0
- `npm` (or `yarn`) version: 5.6.0

A couple widgets that I created seem to have issues stalling when they job runs and gets any type of connection error. Both my gocomic and weather widget pull data from outside sources and if for any instance the job errors trying to get the remote data the job stalls and never runs again unless I restart the entire service, Which lately has been almost every day. Both are simple widgets and can be found in my atlasboard package here - https://github.com/gonzoinc/atlasboard-gonzo-packages

Below is a snippet from the logs when the job connection errors occur and the job never runs again.

```
[dashboard: General] [job: wunderground] 07:40:04.04 <error> executed with errors: bad statusCode: 504 from undefined (in Scheduler.js:21)
Call Stack:
    at handleError (/usr/local/console/builds/1.7.0/node_modules/status-board/lib/Scheduler.js:21:37)
    at jobCallback (/usr/local/console/builds/1.7.0/node_modules/status-board/lib/Scheduler.js:63:21)
    at /usr/local/console/builds/1.7.0/packages/default/jobs/wunderground/wunderground.js:47:13
    at JSON.queryRequest (/usr/local/console/builds/1.7.0/node_modules/status-board/lib/job-dependencies/easyRequest/dependency.js:33:17)
    at Request.request [as _callback] (/usr/local/console/builds/1.7.0/node_modules/status-board/lib/job-dependencies/easyRequest/dependency.js:16:9)
    at Request.self.callback (/usr/local/console/builds/1.7.0/node_modules/request/request.js:186:22)
    at Request.emit (events.js:127:13)
    at Request.<anonymous> (/usr/local/console/builds/1.7.0/node_modules/request/request.js:1163:10)
    
    
Error connecting to url: http://www.gocomics.com/bloomcounty/2018/4/10 { Error: read ECONNRESET
at TCP.onread (net.js:619:25) errno: 'ECONNRESET', code: 'ECONNRESET', syscall: 'read' }
```

The Weather widget pulls json data from Weather Underground and every now and then the feed fails, Which is fine since the board caches the data and will only refresh it when it has new data. the go comics is the same, if it fails to scrape the page to get the latest comic it just continues to use the old one. the Comic is the most noticeable cause I have them set to rotate every couple of min so if I see the same one over and over again I know the job stalled and it's not trying to pull anymore.

Attempts to capture and suppress the error to allow the job to continue failed due to how the job is designed.  I'd rather see the job continue to run even if it fails and either send some sort of notification (I.E. email, message on the board, etc) or to allow a limit for job failure that could be set before the job would die completely.

Thoughts on this?

**Comment by jameswlane** [comment]
@gonzoinc I will see if I can get time to dig into this on the weekend.

If you would like to contribute to the project, if you come up with a solution to resolve it PR's are welcome to the project.

**Comment by jameswlane** [comment]
@gonzoinc, first of all, I want to thank you for all your help the last few months with working the bugs out of Status Board.
I looked at the widgets you made and they are super cool.

Looking into the problem you're having further there is a way to retry on an error a number of times.

```
"retryOnErrorTimes": 5
```

See here: https://github.com/jameswlane/status-board/blob/master/src/Scheduler.ts#L51-L59

You can add it to your job config like so:

```
"gocomics": {
 "retryOnErrorTimes": 5
 "interval": 60000,
 "comic": "calvinandhobbes"
}
```

**Comment by claytoncountryman** [comment]
Hi, I recently ran into this issue too. 

The real issue is that the job needs to use the jobCallback function passed into the onRun function. 

If you look at the scheduler code (Scheduler.js) line 75 you will see that if the callback is not used then nothing happens to re-schedule the task, it just vanishes. 

The Job needs to either not handle its own error (since this would be caught by the scheduler) or the Job needs to pass back an error into the jobCallback function. The jobCallback function handles re-scheduling. 

Looking at your code I see that you handle error by a consol.log which means the jobCallback receives nothing so the scheduler has no idea what to do or how to proceed.

For your gocomics.js do the following

if (!error) {
                var $ = cheerio.load(body);
                var src = $('div div.comic__container div.comic__image.js-comic-swipe-target a picture img').attr('src');

                jobCallback(null, {imageSrc: src});
   
            } else {
                console.error('Error connecting to url:', url, error);
                jobCallback(`<error message>`);//add this or something like it
            }

**Comment by jameswlane** [comment]
Nice find @claytoncountryman I didn't think about that. 👍

**Comment by gonzoinc** [comment]
@claytoncountryman Yep, that did the trick. Once you pointed it out I saw what was going on.  When the request method fails I was only logging it, I wasn't passing the error to the jobCallBack to be processed properly. It's what happens when you overthink a simple problem

@jameswlane I used the retryOnErrorTimes in the config on a different widget and found that if it hits the error limit it will show the yellow Triangle default error on the dashboard and not use what it had cached. I'd rather it use the cached comic in this case so that at least something is showing instead of an error. But I do believe that the retryOnErrorTimes works in conjunction with the jobCallBack being properly passed into the scheduler to handle the retry attempts.

As for PR's, still working out details as to if I'm able to contribute.  I.E. work related.

Thanks again all.
